<div style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=4.5>
		<div align=center>
			<font face="IranNastaliq" size=30>
				<p></p>
				<p></p>
به نام خدا
			</font>
			<br />
			<br />
			<font color=#FF7983>
دانشگاه تهران - دانشکده مهندسی برق و کامپیوتر
            </font>
            <br />
			<font color=midblue>
هوش مصنوعی
            </font>
			<br />
           <font color=#738a9b> 
پاییز ۱۴۰۱
               </div>
		<hr/>
		<font color='red' size=6>
			<div align=center>
پروژه یک
            </div>
		</font>
        <br />
		<div align=center>
بهیاد زرنقی 
            <br />
            شماره دانشجویی: ۸۱۰۶۹۸۲۴۹
        </div>
		<hr />
		<style type="text/css" scoped>
        p{
        border: 1px solid #a2a9b1;background-color: #f8f9fa;display: inline-block;
        };
        </style>            
		
</div>

In [110]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time

At first we read the given text file.

In [278]:
input1=open('input3.txt','r')

The inorder to organize data in the input and convert them to int type parameters, we impose some functions and operators to create an ideal list format file

In [279]:
data = input1.readlines()
input1.close()
modified = list(map(str.strip, data))
print(modified)

['13 16', '1 13', '13 11', '11 7', '7 2', '2 6', '6 12', '12 5', '5 9', '1 4', '2 3', '3 10', '7 8', '10 5', '11 10', '9 2', '9 4', '2', '2 1', '3', '10 3 5 10 3', '4 3 3 11 2', '11 3 1 12 3', '13']


In [281]:
def split_nums(s):
    return  [int(x) for x in s.split()]

In [308]:
list_of_nums = []
for m in modified:
    list_of_nums.append(split_nums(m))

Here is the final list file:

In [307]:
print(list_of_nums)

[[13, 16], [1, 13], [13, 11], [11, 7], [7, 2], [2, 6], [6, 12], [12, 5], [5, 9], [1, 4], [2, 3], [3, 10], [7, 8], [10, 5], [11, 10], [9, 2], [9, 4], [2], [2, 1], [3], [10, 3, 5, 10, 3], [4, 3, 3, 11, 2], [11, 3, 1, 12, 3], [13]]


In the class named Graph we have interpreted different properties of a graph tree including node name, connections between nodes, cooks & recepies and Impassable nodes which are needed in the search procedure

In [309]:
class Graph():
    def __init__(self):
        self.Node()
        self.Connections()
        self.Impassables()
        self.Cooks()
        self.Start()
        
    def Node(self):
        self.name=np.zeros([1,list_of_nums[0][0]]).astype('int32')
        for i in range(list_of_nums[0][0]):
            self.name[0,i]=i+1
        return self.name 
    
    def Connections(self):
        self.cnct=np.zeros([list_of_nums[0][1]+5,list_of_nums[0][0]]).astype('int32')
        self.cnct[0][:]=np.linspace(1,list_of_nums[0][0],list_of_nums[0][0])
        for j in range (list_of_nums[0][0]):
            for i in range (list_of_nums[0][1]):
                if (list_of_nums[i+1][0]==j+1):
                    self.cnct[i+1][j]=list_of_nums[i+1][1]
                if (list_of_nums[i+1][1]==j+1):
                    self.cnct[i+1][j]=list_of_nums[i+1][0]
        return self.cnct
    
    def Impassables(self):
        self.imp=np.zeros(list_of_nums[list_of_nums[0][1]+1]).astype('int32')
        for i in range(list_of_nums[list_of_nums[0][1]+1][0]):
            self.imp[i]=list_of_nums[list_of_nums[0][1]+2][i]
        return self.imp
    
    def Cooks(self):
        self.cks = list_of_nums[list_of_nums[0][1]+3][0]
        self.ck = []
        self.recepies = {}
        self.RCP=[]
        self.rcps=0
        for i in range(1, self.cks+1):
            self.ck.append(list_of_nums[list_of_nums[0][1]+3+i][0])
            self.recepies[self.ck[i-1]] = list_of_nums[list_of_nums[0][1]+3+i][2:]
            for r in list_of_nums[list_of_nums[0][1]+3+i][2:]:
                self.RCP.append(r)

            self.rcps+=list_of_nums[list_of_nums[0][1]+3+i][1]
        return self.recepies
    def Start(self):
        self.strt=list_of_nums[list_of_nums[0][1]+self.cks+4][0]
        return self.strt

In [311]:
vrtx = Graph()

Also the Node parameter has been defined exclussively inorder to explain items like paths,Father of the current node, path , hueristic and... to perform required serach algoritms in further steps.

In [398]:
class Node():
    def __init__(self, name, path, father, isRoot:bool=False):
        self.name = name
        self.isRoot = isRoot
        self.father = father
        if isRoot:
            self.path = str(name)
            self.cost = 0
            self.Lpath = [int(name)]
            self.heuristic=vrtx.cks+vrtx.rcps
            self.f= self.cost+self.heuristic
            self.f1= self.cost+0.9*self.heuristic
            self.f2= self.cost+0.8*self.heuristic
        else:
            self.path = path + ', ' + str(name)
            self.depth = self.path.count(', ')
            self.cost = self.depth
            self.Lpath=[int(x) for x in self.path.split(', ')]
            self.h1=len(list(set(vrtx.ck) & set(self.Lpath)))
            self.h2=len(list(set(vrtx.RCP) & set(self.Lpath)))
            self.heuristic=vrtx.cks+vrtx.rcps-(self.h1+self.h2)
            self.f= self.cost+self.heuristic
            self.f1= self.cost+8*self.heuristic
            self.f2= self.cost+80*self.heuristic
        
        
    def __str__(self):
        return self.path
    def __eq__(self, other):
        if other is not Node:
            return False
        return self.path == other.path
    def __hash__(self):
        return hash(self.path)

In [399]:
def is_goal(node):
    global vrtx
    rpath = node.path.split(', ')
    path = rpath.copy()
    rpath.reverse()
    path = list(map(int, path))
    rpath = list(map(int, rpath))
    for c in vrtx.ck:
        for r in vrtx.recepies[c]:
            if r not in path:
                return False
    for c in vrtx.ck:
        if c not in path:
            return False
    for c in vrtx.ck:
        for r in vrtx.recepies[c]:
            idx_c = len(rpath) - rpath.index(c) - 1
            idx_r = path.index(r)
            if idx_r > idx_c:
                return False
    return True

# BFS Search Alghoritm

In [321]:
from collections import deque
i=0
def BFS(class_type):
    count = 0
    G=class_type
    Queue= deque([])
    Explored=set()
    root = Node(G.strt, "", None, True)
    Queue.append(root)
    Explored.add(root)
    Impassables=list()
    counter=0

    while Queue:          # Creating loop to visit each node
        m = Queue.popleft() 
        if is_goal(m):
            print('Number of states visited:',counter)
            print('Path:', m.path)
            return m
        for neighbour in G.cnct[1: , m.name-1]:
            if neighbour != 0:
                neighbour_node = Node(neighbour, m.path, m.name)
                if neighbour_node not in Explored and neighbour != m.father:
                    counter+=1
                    Explored.add(neighbour_node)
                    Queue.append(neighbour_node)
#                     print(m.path)
                    if neighbour in G.imp and neighbour in m.Lpath:
                        Impassables.append(neighbour)
                        c=Impassables.count(neighbour)
                        for j in range(c-1):
                           Explored.add(None)
                
    return False

In [400]:
%%time
(BFS(Graph()))


Number of states visited: 1412
Path: 13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4
Wall time: 127 ms


In [342]:
Impassables=list()
exp=list()
def DLS(class_type, node, depth_limit):
    G=class_type
    Explored=set()
    if is_goal(node):
        return node
    if depth_limit==0:
        return 0
    
    flag = False
    for neighbour in G.cnct[1: , node.name-1]:
        if neighbour != 0 and neighbour!=node.father:
            exp.append(neighbour)
#                 print(neighbour)
            neighbour_node = Node(neighbour, node.path, node.name)
#             print(neighbour_node.path)
            if neighbour in G.imp and neighbour in neighbour_node.Lpath:
                    Impassables.append(neighbour)
                    c=Impassables.count(neighbour)
                    for j in range(c-1):
                        Explored.add(None)
            res = DLS(class_type, neighbour_node, depth_limit-1)
            if res == 0:
                flag = True
            elif res != -1:
                return res
            
    if flag:
        return 0
    else:
        return -1
                    
    
def IDS(class_type):
    root = Node(vrtx.strt, "", None, True)
    for depth_limit in (range(20)):
        result = DLS(class_type, root, depth_limit)
        if type(result) == Node:
            print('Number of states visited:', len(exp))
            print('Path:', result)
            break

In [343]:
%%time
IDS(Graph())

Number of states visited: 2455
Path: 13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4
Wall time: 172 ms


# A*

as a heuristic we chose the remaining number of Cooks and Recepies which are not visited in the current path. Moreover, each path between two nodes has a cost of one.

In [316]:
from collections import deque
i=0
def A_Star(class_type):
    G=class_type
    Queue= deque([])
    Queue1= deque([])
    Explored=set()
    root = Node(G.strt, "", None, True)
    Queue.append(root)
    Queue1.append(root)
    Explored.add(root)
    neighbour_nodes=list()
    Impassables=list()
    counter=0


    while Queue:          # Creating loop to visit each node
        m = Queue1.pop()
#         print(m.path)

        if is_goal(m):
            print('Number of states visited:',counter)
            print('Path:', m.path)
            return m
        for neighbour in G.cnct[1: , m.name-1]:
            if neighbour != 0:
                neighbour_node= Node(neighbour, m.path, m.name)
                neighbour_nodes.append(neighbour_node)
                if neighbour_node not in Explored and neighbour != m.father:
                    Explored.add(neighbour_node)
                    Queue.append(neighbour_node)
                    counter+=1

#                     print(m.path)
#                     print(m.f)
            if neighbour in G.imp and neighbour in m.Lpath:
                Impassables.append(neighbour)
                c=Impassables.count(neighbour)
                for j in range(c-1):
                    Explored.add(None)
                    
        Min=Queue[0].f
        for n in Queue:
            if n.f<=Min:
                Min=n.f
                best=n
        if best in Queue and best.f==Min:        
            Queue.remove(best)        
            Queue1.append(best)                           
    
    return False                
                    
#                     print(m.path)
               

In [391]:
%%time
(A_Star(Graph()))

Number of states visited: 1135
Path: 13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4
Wall time: 141 ms


In [266]:
from collections import deque
i=0
def A_star1(class_type):
    G=class_type
    Queue= deque([])
    Queue1= deque([])
    Explored=set()
    root = Node(G.strt, "", None, True)
    Queue.append(root)
    Queue1.append(root)
    Explored.add(root)
    neighbour_nodes=list()
    Impassables=list()
    counter=0


    while Queue:          # Creating loop to visit each node
        m = Queue1.pop()
#         print(m.path)

        if is_goal(m):
            print('Number of states visited:',counter)
            print('Path:', m.path)
            return m
        for neighbour in G.cnct[1: , m.name-1]:
            if neighbour != 0:
                neighbour_node= Node(neighbour, m.path, m.name)
                neighbour_nodes.append(neighbour_node)
                if neighbour_node not in Explored and neighbour != m.father:
                    Explored.add(neighbour_node)
                    Queue.append(neighbour_node)
                    counter+=1


            if neighbour in G.imp and neighbour in m.Lpath:
                Impassables.append(neighbour)
                c=Impassables.count(neighbour)
                for j in range(c-1):
                    Explored.add(None)
                    
        Min=Queue[0].f1
        for n in Queue:
            if n.f1<=Min:
                Min=n.f1
                best=n
        if best in Queue and best.f1==Min:        
            Queue.remove(best)        
            Queue1.append(best)                           
    
    return False                
                    
#                     print(m.path)
               

In [394]:
%%time
(A_star1(Graph()))

Number of states visited: 27
Path: 13, 11, 10, 5, 12, 6, 2, 3, 10, 5, 9, 4, 1, 13, 11
Wall time: 7 ms


In [270]:
from collections import deque
i=0
def A_star2(class_type):
    G=class_type
    Queue= deque([])
    Queue1= deque([])
    Explored=set()
    root = Node(G.strt, "", None, True)
    Queue.append(root)
    Queue1.append(root)
    Explored.add(root)
    neighbour_nodes=list()
    Impassables=list()
    counter=0


    while Queue:          # Creating loop to visit each node
        m = Queue1.pop()
#         print(m.path)

        if is_goal(m):
            print('Number of states visited:',counter)
            print('Path:', m.path)
            return m
        for neighbour in G.cnct[1: , m.name-1]:
            if neighbour != 0:
                neighbour_node= Node(neighbour, m.path, m.name)
                neighbour_nodes.append(neighbour_node)
                if neighbour_node not in Explored and neighbour != m.father:
                    Explored.add(neighbour_node)
                    Queue.append(neighbour_node)
                    counter+=1


            if neighbour in G.imp and neighbour in m.Lpath:
                Impassables.append(neighbour)
                c=Impassables.count(neighbour)
                for j in range(c-1):
                    Explored.add(None)
                    
        Min=Queue[0].f2
        for n in Queue:
            if n.f2<=Min:
                Min=n.f2
                best=n
        if best in Queue and best.f2==Min:        
            Queue.remove(best)        
            Queue1.append(best)                           
    
    return False 

In [397]:
%%time
(A_star2(Graph()))

Number of states visited: 27
Path: 13, 11, 10, 5, 12, 6, 2, 3, 10, 5, 9, 4, 1, 13, 11
Wall time: 4 ms


### Results: First TestCase

| Search Algoritm | Number of Explored States | Time | Path |
| --- | --- | --- | --- |
| BFS | 500 |  46ms |  1, 3, 4, 5, 7, 10, 11, 9, 8  |
| IDS | 529 | 42ms |  1, 3, 4, 5, 7, 10, 11, 9, 8  |
| A* | 267 | 44ms |  1, 3, 4, 5, 7, 10, 11, 9, 8 |
| WeightedA* 1 | 23 | 3ms |  1, 3, 10, 11, 9, 8, 7, 5, 4, 8  | 
| WeightedA* 2 | 23 | 3ms |  1, 3, 10, 11, 9, 8, 7, 5, 4, 8   |


### Results: Second TestCase

| Search Algoritm | Number of Explored States | Time | Path |
| --- | --- | --- | --- |
| BFS | 752 |  63ms |  9, 10, 6, 4, 12, 3, 7, 5, 8  |
| IDS | 890 | 57ms |  9, 10, 6, 4, 12, 3, 7, 5, 8  |
| A* | 114 | 19ms |  9, 10, 6, 4, 12, 3, 7, 5, 8  |
| WeightedA* 1 | 108 | 18ms |  9, 10, 8, 5, 7, 3, 12, 4, 6, 10, 8, 5, 7   | 
| WeightedA* 2 | 108 | 18ms |  9, 10, 8, 5, 7, 3, 12, 4, 6, 10, 8, 5, 7   |


### Results: Third TestCase

| Search Algoritm | Number of Explored States | Time | Path |
| --- | --- | --- | --- |
| BFS | 1412 |  148ms |  13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4  |
| IDS | 2455 | 140ms |  13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4  |
| A* | 1135 | 134ms |  13, 1, 4, 9, 5, 12, 6, 2, 3, 10, 11, 13, 1, 4  | -->
| WeightedA* 1 | 27 | 5.6ms |  13, 11, 10, 5, 12, 6, 2, 3, 10, 5, 9, 4, 1, 13, 11  | 
| WeightedA* 2 | 27 | 5ms |  13, 11, 10, 5, 12, 6, 2, 3, 10, 5, 9, 4, 1, 13, 11   |
